File: Day114_2025/Q114_Leetcode.py
from typing import List


class Solution:
    def countCompleteSubarrays(self, nums: List[int]) -> int:
        from collections import defaultdict

        n = len(nums)
        res = 0
        distinct_count = len(set(nums))
        cnt = defaultdict(int)
        right = 0

        for left in range(n):
            if left > 0:
                cnt[nums[left - 1]] -= 1
                if cnt[nums[left - 1]] == 0:
                    del cnt[nums[left - 1]]

            while right < n and len(cnt) < distinct_count:
                cnt[nums[right]] += 1
                right += 1

            if len(cnt) == distinct_count:
                res += (n - right + 1)

        return res

File: Day114_2025/Q114_Leetcode.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class Q114_Leetcode {
    public int countCompleteSubarrays(int[] nums) {
        int n = nums.length;
        int res = 0;
        
        // Find number of distinct elements in the array
        Set<Integer> distinctSet = new HashSet<>();
        for (int num : nums) distinctSet.add(num);
        int distinctCount = distinctSet.size();

        Map<Integer, Integer> countMap = new HashMap<>();
        int right = 0;

        for (int left = 0; left < n; left++) {
            if (left > 0) {
                int remove = nums[left - 1];
                countMap.put(remove, countMap.get(remove) - 1);
                if (countMap.get(remove) == 0) {
                    countMap.remove(remove);
                }
            }

            while (right < n && countMap.size() < distinctCount) {
                countMap.put(nums[right], countMap.getOrDefault(nums[right], 0) + 1);
                right++;
            }

            if (countMap.size() == distinctCount) {
                res += (n - right + 1);
            }
        }

        return res;
    }
}
